### PR TITLE
add circle loss and test scripts

### PR DESCRIPTION
**Pull request by AlenUbuntu**
Hello, all:

I have added a circle loss into the repo. You may review the change at losses/circle_loss.py
A test script is also provided at tests/losses/test_circle_loss.py. 

For your information, here are the training and evaluation results on CIFAR 100 dataset. I simply compare to the triplet loss with margin = 0.01. In general, circle loss converges faster and better than triplet loss. The default hyper-parameter settings in circle loss are also given in the source script file.

Triplet Loss:

Training 100/100 - loss: 0.0767
Evaluating: 100/100
Epoch: 0 -- 5NN Acc: 0.2428
Training 100/100 - loss: 0.0702
Evaluating: 100/100
Epoch: 1 -- 5NN Acc: 0.2654
Training 100/100 - loss: 0.0597
Evaluating: 100/100
Epoch: 2 -- 5NN Acc: 0.2753
Training 100/100 - loss: 0.0621
Evaluating: 100/100
Epoch: 3 -- 5NN Acc: 0.2858
Training 100/100 - loss: 0.0466
Evaluating: 100/100
Epoch: 4 --  5NN Acc: 0.2968
Training 100/100 - loss: 0.0439
Evaluating: 100/100
Epoch: 5 -- 5NN Acc: 0.2977
Training 100/100 - loss: 0.0445
Evaluating: 100/100
Epoch: 6 -- 5NN Acc: 0.3076
Training 100/100 - loss: 0.0381
Evaluating: 100/100
Epoch: 7 -- 5NN Acc: 0.3232
Training 100/100 - loss: 0.0328
Evaluating: 100/100
Epoch: 8 -- 5NN Acc: 0.3221
Training 100/100 - loss: 0.0421
Evaluating: 100/100
Epoch: 9 -- 5NN Acc: 0.3335
Training 100/100 - loss: 0.0360
Evaluating: 100/100
Epoch: 10 -- 5NN Acc: 0.3432
Training 100/100 - loss: 0.0266
Evaluating: 100/100
Epoch: 11 -- 5NN Acc: 0.3476
Training 100/100 - loss: 0.0328
Evaluating: 100/100
Epoch: 12 -- 5NN Acc: 0.3498
Training 100/100 - loss: 0.0271
Evaluating: 100/100
Epoch: 13 -- 5NN Acc: 0.3514
Training 100/100 - loss: 0.0297
Evaluating: 100/100
Epoch: 14 -- 5NN Acc: 0.3585
Training 100/100 - loss: 0.0322
Evaluating: 100/100
Epoch: 15 -- 5NN Acc: 0.3754
Training 100/100 - loss: 0.0276
Evaluating: 100/100
Epoch: 16 -- 5NN Acc: 0.3726
Training 100/100 - loss: 0.0266
Evaluating: 100/100
Epoch: 17 -- 5NN Acc: 0.3881
Training 100/100 - loss: 0.0283
Evaluating: 100/100
Epoch: 18 --  5NN Acc: 0.3772
Training 100/100 - loss: 0.0304
Evaluating: 100/100
Epoch: 19 -- 5NN Acc: 0.3872

Circle Loss:

Training 100/100 - loss: 50.5098
Evaluating: 100/100
Epoch: 0 -- 5NN Acc: 0.2435
Training 100/100 - loss: 50.4540
Evaluating: 100/100
Epoch: 1 -- 5NN Acc: 0.2647
Training 100/100 - loss: 46.3977
Evaluating: 100/100
Epoch: 2 -- 5NN Acc: 0.2788
Training 100/100 - loss: 48.3472
Evaluating: 100/100
Epoch: 3 -- 5NN Acc: 0.3060
Training 100/100 - loss: 47.5886
Evaluating: 100/100
Epoch: 4 -- 5NN Acc: 0.3171
Training 100/100 - loss: 47.3185
Evaluating: 100/100
Epoch: 5 -- 5NN Acc: 0.3280
Training 100/100 - loss: 47.0622
Evaluating: 100/100
Epoch: 6 -- 5NN Acc: 0.3423
Training 100/100 - loss: 46.4507
Evaluating: 100/100
Epoch: 7 -- 5NN Acc: 0.3550
Training 100/100 - loss: 46.5572
Evaluating: 100/100
Epoch: 8 -- 5NN Acc: 0.3653
Training 100/100 - loss: 45.9522
/Evaluating: 100/100
Epoch: 9 -- 5NN Acc: 0.3674
Training 100/100 - loss: 44.3513
Evaluating: 100/100
Epoch: 10 -- 5NN Acc: 0.3801
Training 100/100 - loss: 46.0787
Evaluating: 100/100
Epoch: 11 -- 5NN Acc: 0.3866
Training 100/100 - loss: 48.0457
Evaluating: 100/100
Epoch: 12 -- 5NN Acc: 0.3936
Training 100/100 - loss: 43.3233
Evaluating: 100/100
Epoch: 13 -- 5NN Acc: 0.3954
Training 100/100 - loss: 46.2095
Evaluating: 100/100
Epoch: 14 -- 5NN Acc: 0.4025
Training 100/100 - loss: 42.1961
Evaluating: 100/100
Epoch: 15 -- 5NN Acc: 0.4128
Training 100/100 - loss: 40.3038
Evaluating: 100/100
Epoch: 16 -- 5NN Acc: 0.4226
Training 100/100 - loss: 47.8131
Evaluating: 100/100
Epoch: 17 -- 5NN Acc: 0.4317
Training 100/100 - loss: 45.3652
Evaluating: 100/100
Epoch: 18 -- 5NN Acc: 0.4358
Training 100/100 - loss: 43.9636
Evaluating: 100/100
Epoch: 19 -- 5NN Acc: 0.4385

Please let me know if you find any issues. Thanks.

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
@AlenUbuntu Thanks a lot for doing this! I just made some minor changes which you can see [here](https://github.com/KevinMusgrave/pytorch-metric-learning/commit/3326921c5a8480b49d0a5e2cdc45e31eb04cccc6) and [here](https://github.com/KevinMusgrave/pytorch-metric-learning/commit/c6e6a18f4a4761b596cd05770242238881b0bd3c). 

Your test looks very useful, but I removed it because the tests folder is meant for unit-tests, i.e. tests for individual components. So a test for CircleLoss would be something like verifying the loss value for some handcrafted embedding that we pass in. 

I do like the way you calculated accuracy (predicting label by majority vote of knn), so I think it would be good to add that functionality to the default AccuracyCalculator.

**Comment by AlenUbuntu** [comment]
Hi, Kevin:

Thanks for improving my code. Please include this loss in your ReadMe file
accordingly. I forgot to do that.

Best

Yang

**Comment by KevinMusgrave** [comment]
@AlenUbuntu Done. I updated the losses and contributors section of the readme, and updated the documentation
